src/cfml/system/endpoints/S3.cfc

```diff
@@ -0,0 +1,193 @@
+        var host = 's3.#bucketRegion#.amazonaws.com';
+        var path = encodeUrl('/#bucket#/#objectKey#', false);
+
+        // query string
+        var qs = {
+            'X-Amz-Algorithm': 'AWS4-HMAC-SHA256',
+            'X-Amz-Credential': awsSettings.awsKey & '/' & isoTime.left( 8 ) & '/' & bucketRegion & '/s3/aws4_request',
+            'X-Amz-Date': isoTime,
+            'X-Amz-Expires': 300,
+            'X-Amz-SignedHeaders': 'host'
+        };
+        if (awsSettings.sessionToken.len()) {
+            qs['X-Amz-Security-Token'] = awsSettings.sessionToken;
+        }
+        qs = qs.keyArray()
+            .sort('text')
+            .reduce((r, key) => r.listAppend('#key#=#encodeUrl(qs[key])#', '&'), '');
+
+        // canonical request
+        var canonicalRequest = [
+            'GET',
+            path,
+            qs,
+            'host:#host#',
+            '',
+            'host',
+            'UNSIGNED-PAYLOAD'
+        ].toList(chr(10));
+
+        // string to sign
+        var stringtoSign = [
+            'AWS4-HMAC-SHA256',
+            isoTime,
+            isoTime.left( 8 ) & '/' & bucketRegion & '/s3/aws4_request',
+            hash(canonicalRequest, 'SHA-256').lcase()
+        ].toList(chr(10));
+
+        // signature
+        var signingKey = binaryDecode(hmac(isoTime.left(8), 'AWS4' & awsSettings.awsSecretKey, 'hmacSHA256', 'utf-8'), 'hex');
+        signingKey = binaryDecode(hmac(bucketRegion, signingKey, 'hmacSHA256', 'utf-8'), 'hex');
+        signingKey = binaryDecode(hmac('s3', signingKey, 'hmacSHA256', 'utf-8'), 'hex');
+        signingKey = binaryDecode(hmac('aws4_request', signingKey, 'hmacSHA256', 'utf-8'), 'hex');
+        var signature = hmac(stringToSign, signingKey, 'hmacSHA256', 'utf-8').lcase();
+
+        qs &= '&X-Amz-Signature=' & signature;
+        return '//' & host & path & '?' & qs;
+    }
+
+    private function resolveBucketRegion(bucket, defaultRegion) {
+        var req = '';
+        cfhttp(url='https://s3.#defaultRegion#.amazonaws.com', method='HEAD', result='req', redirect=false) {
+            cfhttpparam(type='header', name='Host', value='#bucket#.s3.amazonaws.com');
+        }
+        return req.responseheader['x-amz-bucket-region'];
+    }
+
+    private function resolveAwsSettings() {
+        var settings = {
+            awsKey: systemSettings.getSystemSetting('AWS_ACCESS_KEY_ID', ''),
+            awsSecretKey: systemSettings.getSystemSetting('AWS_SECRET_ACCESS_KEY', ''),
+            sessionToken: systemSettings.getSystemSetting('AWS_SESSION_TOKEN', ''),
+            defaultRegion: systemSettings.getSystemSetting('AWS_DEFAULT_REGION', ''),
+            profile: systemSettings.getSystemSetting('AWS_PROFILE', 'default'),
+            configFile: systemSettings.getSystemSetting('AWS_CONFIG_FILE', '~/.aws/config'),
+            credentialsFile: systemSettings.getSystemSetting('AWS_SHARED_CREDENTIALS_FILE', '~/.aws/credentials')
```

Comment:
I'm not totally familiar with how AWS is normally configured, but does it make sense to default these to config settings under `endpoints.s3.etc`?  Just to give people the ability to set this config inside of CommandBox is env vars weren't their thing.